src/provider/proxyprovider.cpp

```diff
@@ -0,0 +1,86 @@
+#include "provider/proxyprovider.h"
+#include "apiresponse.h"
+#include "httpwrapper.h"
+
+namespace internal
+{
+    bool isRequestSuccessful(ErdGenericApiResponse const &response)
+    {
+        std::string const error = response.getError();
+        std::string const code = response.getCode();
+
+        return error.empty() && (code.find("success")!= std::string::npos);
+    }
+
+    nlohmann::json getDataIfValid(wrapper::http::Result const &res)
+    {
+        if (res.error)
+        {
+            throw std::runtime_error(res.statusMessage);
+        }
+
+        ErdGenericApiResponse wrappedResponse(res.body);
+
+        if (res.status != STATUS_CODE_OK || !isRequestSuccessful(wrappedResponse))
+        {
+            throw std::runtime_error(ERROR_MSG_HTTP_REQUEST_FAILED + res.statusMessage + ". " +
+                                     ERROR_MSG_REASON + wrappedResponse.getError());
+        }
+
+        return wrappedResponse.getData<nlohmann::json>();
+    }
+
+}
+
+ProxyProvider::ProxyProvider(std::string url):
+    m_url(std::move(url)) {}
+
+Account ProxyProvider::getAccount(Address const &address)
+{
+    wrapper::http::Client client(m_url);
+    wrapper::http::Result const result = client.get("/address/" + address.getBech32Address());
+
+    auto data = internal::getDataIfValid(result);
+
+    if (!data.contains("account"))
+        throw std::invalid_argument(ERROR_MSG_JSON_KEY_NOT_FOUND + "account");
+    if (!data["account"].contains("balance"))
+        throw std::invalid_argument(ERROR_MSG_JSON_KEY_NOT_FOUND + "balance");
+    if (!data["account"].contains("nonce"))
+        throw std::invalid_argument(ERROR_MSG_JSON_KEY_NOT_FOUND + "nonce");
+
+    std::string const balance = data["account"]["balance"];
+    uint64_t const nonce = data["account"]["nonce"];
+
+    return Account(address, balance, nonce);
+}
+
+TransactionHash ProxyProvider::send(Transaction const &transaction)
+{
+    wrapper::http::Client client(m_url);
+    wrapper::http::Result const result = client.post("/transaction/send", transaction.serialize(), wrapper::http::applicationJson);
+
+    auto data = internal::getDataIfValid(result);
+
+    if (!data.contains("txHash"))
```

Comment:
As above. Doesn't `data["txHash"];` throw a good-enough exception if the key is missing, anyway?

Reply:
As discussed yesterday, we will keep it as it is